README.md
BookingBot - 口罩预约机器人
===
## 软件说明
现在市面上大量抢口罩的脚本,导致我抢不到口罩,本着打不过就打入的原则,自己随手写了个抢口罩的小工具,
效率一般,因为写的急,加上还有别的事,就没在意可维护性,大佬轻喷,不过目测这个工具也
用不了多久,希望疫情能尽早结束。      
药店id就不公布了,作者就收集了2个家附近的药店id,这样会暴露作者所处的地理位置,
拿去用的朋友可以自己分析一下。   
  
**`帮女票家人抢到N次，值得信赖`**  

## 使用说明
**参数说明**:  
<pre>
    -h 参数列表及使用方法
    -conf 指定配置文件
    -v 版本号
</pre>  
配置文件默认目录:**./config/config.json**  
**配置文件配置说明**:  
<pre>
    account:手机号
    password:密码
    placePointId:药店id,
    sleep:抢购间隔(单位:毫秒)
    max:最大抢购重试次数
    companyId:企业(地区、类别)id,这个id是用来区分2.5元和8毛的
    num:抢购个数(最大5)
    grade:线程(同时几个worker帮你抢)
</pre>
## 更新日志

[更新日志](CHANGELOG.md)

## 许可
[LICENSE](LICENSE)

## 作者
[BuTn](https://github.com/kimmosc2)

CHANGELOG.md

## 更新日志
#### 200312.2103.Pro_beta  
 - 重写了软件架构,不再支持手动输入,完全由配置文件制定  
 - 单个配置文件支持多个账号
  
#### 200310_1852
修复了地区的bug

#### 200310_1720_alpha
1.支持其他地区口罩的预约(暂未测试)    
2.支持指定配置文件路径  


cmd/main.go
package main

import (
	config2 "booking-bot/config"
	"booking-bot/flags"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"net/http/cookiejar"
	"os"
	"strconv"
	"strings"
	"time"
	// "github.com/peterh/liner"
)

// TODO HeaderFactory

// 登录表单
type loginForm struct {
	UserForm  `json:"data"`
	CompanyId string `json:"companyid"`
}

// 手机号，密码
type UserForm struct {
	Account  string `json:"account"`
	Password string `json:"password"`
}

// 登录返回信息
type Result struct {
	Data   User   `json:"data"`
	Msg    string `json:"msg"`
	Status int    `json:"status"`
}

// 用户信息(登录后返回的)
type User struct {
	// 市
	Country string `json:"country"`
	// 详细地址
	Address string `json:"address"`
	// 省
	City string `json:"city"`
	// 姓名
	Name string `json:"name"`
	// userId,购买时需要的
	Id string `json:"id"`
}

// 每天id结构体解码
type TodayList struct {
	Data TodayData `json:"data"`
}

type TodayData struct {
	TodayGoodsDtos []TodayGoodsDtos `json:"todayGoodsDtos"`
}

type TodayGoodsDtos struct {
	// 药店名
	PlacepointName string `json:"placepointName"`
	// 口罩id
	Id int `json:"id"`
}

// 抢购返回json
type Response struct {
	Data   string `json:"data"`
	Msg    string `json:"msg"`
	Status int    `json:"status"`
}

var (
	client = new(http.Client)
	// 登录凭证
	token      string
	loginData  = new(loginForm)
	config     = new(config2.Config)
	isLogin    = false
	userInfo   = new(User)
	targetInfo = new(TodayList)
)

// cookie 初始化
func cookieInit() {
	jar, err := cookiejar.New(nil)
	if err != nil {
		log.Fatal("初始化cookie失败:", err.Error())
	}
	client.Jar = jar
}

// 表单初始化
func formInit() {
	loginData.CompanyId = config.CompanyId
}

func parseConfig() {

	open, err := os.Open(flags.ConfigPath)
	if err != nil {
		log.Fatal("open file error:", err.Error())
	}
	b, _ := ioutil.ReadAll(open)
	err = json.Unmarshal(b, &config)
	if err != nil {
		log.Fatal("parse config error:", err.Error())
	}
	if config.Num > 5 {
		log.Fatal("最大抢购数量为5")
	}
	if config.CompanyId == "" {
		log.Fatal("请填写地区id")
	}
}

// 获取当日口罩id, data是药店id, id是购买者id,token是登录token
func fetchId(data, id, token, companyId string) {
	body := `{"companyId":"` + companyId + `","data":"` + data + `","loginid":"` + id + `"}`
	req, err := http.NewRequest(http.MethodPost, "http://reservepc.inca.com.cn:9345/booking/todayList", strings.NewReader(body))
	if err != nil {
		panic(err)
	}
	req.Header.Set("Accept", `application/json, text/plain, */*`)
	req.Header.Set("Accept-Encoding", `gzip, deflate`)
	req.Header.Set("Accept-Language", `zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6`)
	req.Header.Set("Connection", `keep-alive`)
	req.Header.Set("Content-Type", `application/json;charset=UTF-8`)
	req.Header.Set("Referer", `http://reserveyd.inca.com.cn/index?companyid=`+companyId)
	req.Header.Set("User-Agent", `Mozilla/5.0 (iPhone; CPU iPhone OS 10_3_1 like Mac OS X) AppleWebKit/603.1.30 (KHTML, like Gecko) Version/10.0 Mobile/14E304 Safari/602.1`)
	req.Header.Set("token", token)

	for {
		fmt.Println("正在获取药店信息...")
		do, err := client.Do(req)
		if err == nil {
			b, _ := ioutil.ReadAll(do.Body)
			err := json.Unmarshal(b, &targetInfo)
			if err != nil {
				log.Fatal("unmarshal failed:", err.Error())
			}
			if len(targetInfo.Data.TodayGoodsDtos) > 0 && targetInfo.Data.TodayGoodsDtos[0].Id != 0 {
				break
			}
		}
	}
}

func init() {
	flag.Parse()
	file := "./" + "message" + ".txt"
	logFile, err := os.OpenFile(file, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0766)
	if err != nil {
		panic(err)
	}
	log.SetOutput(logFile) // 将文件设置为log输出的文件
	log.SetPrefix("[booking-bot]")
	log.SetFlags(log.LstdFlags | log.Lshortfile | log.LUTC)
}

func main() {
	if flags.Version {
		fmt.Println("BookingBot_200310_1720_alpha")
		os.Exit(0)
	}
	if flags.Help {
		flag.Usage()
		os.Exit(0)
	}

	cookieInit()
	formInit()
	parseConfig()
	fmt.Println("指定的配置文件:", flags.ConfigPath)
	targetInfo.Data.TodayGoodsDtos = make([]TodayGoodsDtos, 1)
	var user UserForm
	user.Account = config.Account
	user.Password = config.Password
	loginData.UserForm = user
	marshal, _ := json.Marshal(loginData)
	// 登录
	for !isLogin {
		Login(string(marshal))
	}
	fetchId(config.PlacePointId, userInfo.Id, token, config.CompanyId)
	fmt.Println("本次预约:", targetInfo.Data.TodayGoodsDtos[0].PlacepointName)
	// 买
	realBuy(strconv.Itoa(config.Num), config.PlacePointId, strconv.Itoa(targetInfo.Data.TodayGoodsDtos[0].Id), config.CompanyId, config.Sleep, config.Max)

}

func realBuy(num string, placePointId string, saleGoodsId, companyId string, sleep, max int) {
	for i := 1; i <= max; i++ {
		fmt.Printf("第%d次尝试抢购:", i)
		// 构造购买请求
		buy := Buy(num, placePointId, saleGoodsId, userInfo.Id, companyId)

		// 发起购买
		resp, err := client.Do(buy)
		if err != nil {
			panic(err)
		}
		b, _ := ioutil.ReadAll(resp.Body)
		response := new(Response)
		json.Unmarshal(b, &response)
		fmt.Println(response.Msg) // 抢购结果
		if response.Status == 200 {
			break
		}
		resp.Body.Close()
		time.Sleep(time.Millisecond * time.Duration(sleep))
	}
	fmt.Println("完成,抢购结果可扫描二维码进行查看,程序将在30秒后结束")
	time.Sleep(time.Second * 30)
}

// 登录
func Login(body string) {
	url := `http://reservepc.inca.com.cn:9345/login/auth`
	request, err := http.NewRequest(http.MethodPost, url, strings.NewReader(body))
	if err != nil {
		log.Fatal("登录请求初始化失败:", err.Error())
	}
	request.Header.Set("Connection", `keep-alive`)
	request.Header.Set("Accept", `application/json, text/plain, */*`)
	request.Header.Set("User-Agent", `Mozilla/5.0 (iPhone; CPU iPhone OS 10_3_1 like Mac OS X) AppleWebKit/603.1.30 (KHTML, like Gecko) Version/10.0 Mobile/14E304 Safari/602.1`)
	request.Header.Set("Content-Type", `application/json;charset=UTF-8`)
	request.Header.Set("Accept-Encoding", `gzip, deflate`)
	request.Header.Set("Accept-Language", `zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6`)
	resp, err := client.Do(request)
	if err != nil {
		log.Fatal("在执行登录时出现了一个错误:", err.Error())
	}
	defer resp.Body.Close()
	b, _ := ioutil.ReadAll(resp.Body)

	result := new(Result)
	json.Unmarshal(b, &result)

	if result.Status != 200 {
		fmt.Println("登陆失败:", result.Msg)
		isLogin = false
	} else {
		isLogin = true
		fmt.Printf("您好:%s\n", result.Data.Name)
		userInfo = &result.Data

		token = resp.Header.Get("Token")
	}

}

// 购买
func Buy(num, placePointId, saleGoodsId, userId, companyId string) *http.Request {
	body := `{"companyId":"` + companyId + `","data":{"address":"","deliverState":"0","goodsQty":` + num + `,"zdGoodsName":"一次性医用口罩","placepointId":"` + placePointId + `","saleGoodsId":` + saleGoodsId + `,"userId":"` + userId + `"}}`
	request, err := http.NewRequest(http.MethodPost, "http://reservepc.inca.com.cn:9345/booking/save", strings.NewReader(body))
	if err != nil {
		panic(err)
	}
	request.Header.Set("token", token)
	request.Header.Set("Connection", `keep-alive`)
	request.Header.Set("Accept", `application/json, text/plain, */*`)
	request.Header.Set("User-Agent", `Mozilla/5.0 (iPhone; CPU iPhone OS 10_3_1 like Mac OS X) AppleWebKit/603.1.30 (KHTML, like Gecko) Version/10.0 Mobile/14E304 Safari/602.1`)
	request.Header.Set("Content-Type", `application/json;charset=UTF-8`)
	request.Header.Set("Origin", `http://reserveyd.inca.com.cn`)
	request.Header.Set("Referer", `http://reserveyd.inca.com.cn/index?companyid=0ac83ca4260a4549acfb383a14665783`)
	request.Header.Set("Accept-Encoding", `gzip, deflate`)
	request.Header.Set("Accept-Language", `zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6`)

	return request
}


cmd/booking.go
package main

import (
	"booking-bot/config"
	"booking-bot/datasource"
	"booking-bot/flags"
	"flag"
	"fmt"
	"log"
	"os"
	"sync"
	"time"
)
func init() {
	flag.Parse()
	file := "./" + "message" + ".txt"
	logFile, err := os.OpenFile(file, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0766)
	if err != nil {
		panic(err)
	}
	log.SetOutput(logFile) // 将文件设置为log输出的文件
	log.SetPrefix("[booking-bot]")
	log.SetFlags(log.LstdFlags | log.Lshortfile | log.LUTC)
}
func main() {
	if flags.Version {
		fmt.Println("BookingBot 200312.2103.Pro_beta\nAuthor:BuTn<kimmosc2@163.com>")
		os.Exit(0)
	}
	if flags.Help {
		flag.Usage()
		os.Exit(0)
	}
	users := datasource.NewConfiguration()
	config.LoadConfig(users)
	// 登录区
	loginChan := make(chan struct{}, 1)
	for _, u := range users.Users {
		go u.Login(loginChan)
	}
	for i := len(users.Users); i != 0; i-- {
		<-loginChan
	}
	fmt.Println("所有用户都已装载完毕")

	wgp := new(sync.WaitGroup)
	for _, u := range users.Users {
		wgp.Add(1)
		go u.FetchId(wgp)
	}
	wgp.Wait()

	fmt.Println("完成,抢购结果可扫描二维码进行查看,程序将在30秒后结束")
	time.Sleep(time.Second * 30)
}




datasource/user.go
package datasource

import (
	"booking-bot/util"
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"net/http/cookiejar"
	"strconv"
	"strings"
	"sync"
	"time"
)

func NewConfiguration() *Configuration {
	c := new(Configuration)
	c.Users = make([]*User, 1)
	return c
}

type Configuration struct {
	Users []*User `json:"users"`
}

// 用户信息
type User struct {
	// 账号
	Account string `json:"account"`
	// 密码
	Password string `json:"password"`
	// httpClient
	Client *http.Client
	// 登录状态
	IsLogin bool
	// 市
	Country string `json:"country"`
	// 详细地址
	Address string `json:"address"`
	// 省
	City string `json:"city"`
	// 姓名
	Name string `json:"name"`
	// userId,购买时需要的
	Id string `json:"id"`
	// Token
	Token string `json:"token"`
	// 药店id
	PlacePointId string `json:"placePointId"`
	// 药店名称
	PlacepointName string `json:"placepointName"`
	// 间隔时间
	Sleep int `json:"sleep"`
	// 最大重试次数
	Max int `json:"max"`
	// 预约个数
	Num int `json:"num"`
	// 地区id
	CompanyId string `json:"companyId"`
	// 口罩id,每天变化,提前十分钟获取
	SaleGoodsId int `json:"saleGoodsId"`
	// 口罩名称
	ZgGoodsName string `json:"zdGoodsName"`
	// Goroutine数量
	Grade int `json:"grade"`
	// 登录次数
	loginCount int
	// 预约是否成功
	BookingSuccess bool `json:"success"`
	// 有id
	HasId bool `json:"hasId"`
}

type LoginForm struct {
	User      `json:"data"`
	CompanyId string `json:"companyid"`
}

// 内嵌循环,登录不成功就循环登录
func (u *User) Login(ch chan struct{}) {
	for !u.IsLogin {
		u.loginCount++
		url := `http://reservepc.inca.com.cn:9345/login/auth`
		form := new(LoginForm)
		bindForm(form, u)
		user, err := json.Marshal(form)
		if err != nil {
			log.Fatal("json marshal error:", err)
		}
		request, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(user))
		if err != nil {
			log.Fatal("登录请求初始化失败:", err.Error())
		}
		util.SetAutoRequestHeader(request, "")
		if u.Client == nil {
			u.Client = new(http.Client)
			jar, _ := cookiejar.New(nil)
			u.Client.Jar = jar
		}

		resp, err := u.Client.Do(request)
		if err != nil {
			log.Fatal("在执行登录时出现了一个错误:", err.Error())
		}
		b, _ := ioutil.ReadAll(resp.Body)
		resp.Body.Close()
		r := new(Response)
		r.Data = u
		err = json.Unmarshal(b, &r)
		if err != nil {
			log.Fatal("json unmarshal error:", err)
		}

		if r.Status != 200 {
			if u.loginCount >= 5 {
				fmt.Printf("%v已连续登录失败5次,已取消登录\n", u.Account)
				ch <- struct{}{}
				break
			}
			fmt.Printf("%v登陆失败:%v,正在重试\n", u.Account, r.Msg)
			u.IsLogin = false
		} else {
			u.IsLogin = true
			fmt.Printf("%v登录成功,姓名:%s\n", u.Account, u.Name)
			u.Token = resp.Header.Get("Token")
			ch <- struct{}{}
			break
		}
		time.Sleep(time.Second)
	}
}

func (u *User) FetchId(wgp *sync.WaitGroup) {
	if !u.IsLogin {
		return
	}
	body := `{"companyId":"` + u.CompanyId + `","data":"` + u.PlacePointId + `","loginid":"` + u.Id + `"}`
	req, err := http.NewRequest(http.MethodPost, "http://reservepc.inca.com.cn:9345/booking/todayList", strings.NewReader(body))
	if err != nil {
		log.Fatal("build request error:", err.Error())
	}
	util.SetAutoRequestHeader(req, u.Token)

	for {
		fmt.Printf("%v正在获取药店信息\n", u.Name)
		do, err := u.Client.Do(req)
		if err == nil {
			b, _ := ioutil.ReadAll(do.Body)
			t := new(TodayList)
			t.Data.TodayGoodsDtos = make([]TodayGoodsDtos, 1)
			err := json.Unmarshal(b, &t)
			if err != nil {
				log.Println("unmarshal saleid error:", err.Error())
				return
			}
			if len(t.Data.TodayGoodsDtos) > 0 && t.Data.TodayGoodsDtos[0].Id != 0 {
				u.ZgGoodsName = t.Data.TodayGoodsDtos[0].ZdGoodsName
				u.PlacepointName = t.Data.TodayGoodsDtos[0].PlacepointName
				u.SaleGoodsId = t.Data.TodayGoodsDtos[0].Id
				u.Buy()
				wgp.Done()
				break
			}
		}
		time.Sleep(time.Second)
	}
}

func (u *User) Buy() {
	if !u.IsLogin {
		return
	}
	count := 0
	wg := new(sync.WaitGroup)
	for i := 0; i < u.Grade; i++ {
		wg.Add(1)
		go func(worker int, c int, user *User, w *sync.WaitGroup) {
			for ; c <= user.Max; c++ {
				if user.BookingSuccess {
					w.Done()
					break
				}
				// 构造购买请求
				r := buildBookingRequest(strconv.Itoa(user.Num), user.PlacePointId, strconv.Itoa(user.SaleGoodsId), user.Id, user.CompanyId, user.Token, user.ZgGoodsName)
				realBuy(user, r)
				time.Sleep(time.Millisecond * time.Duration(user.Sleep))
			}
		}(i, count, u, wg)
	}
	wg.Wait()
}

func realBuy(u *User, req *http.Request) {
	resp, err := u.Client.Do(req)
	if err != nil {
		log.Println("request error:", err.Error())
		return
	}
	b, _ := ioutil.ReadAll(resp.Body)
	resp.Body.Close()
	response := new(Response)
	err = json.Unmarshal(b, &response)
	if err != nil {
		log.Println("unmarshal error:", err.Error())
		return
	}

	if response.Status == 200 {
		fmt.Printf("%v抢购成功\n", u.Name)
		log.Printf("%v抢购成功\n", u.Name)
		u.BookingSuccess = true
		return
	}
	fmt.Printf("【%s】 | 【%s】 | 【%s】 | 【%s】  \n", u.Name, u.PlacepointName, u.ZgGoodsName, response.Msg) // 抢购结果
}

func buildBookingRequest(num, placePointId, saleGoodsId, userId, companyId, token, zgGoodsName string) *http.Request {
	body := `{"companyId":"` + companyId + `","data":{"address":"","deliverState":"0","goodsQty":` + num +
		`,"zdGoodsName":"` + zgGoodsName + `","placepointId":"` + placePointId + `","saleGoodsId":` + saleGoodsId +
		`,"userId":"` + userId + `"}}`
	request, err := http.NewRequest(http.MethodPost, "http://reservepc.inca.com.cn:9345/booking/save", strings.NewReader(body))
	if err != nil {
		log.Fatal("build Buy request error:", err.Error())
	}
	util.SetAutoRequestHeader(request, token)
	return request
}

func bindForm(form *LoginForm, u *User) {
	form.CompanyId = u.CompanyId
	form.Account = u.Account
	form.Password = u.Password
}
